# Financial market News - Sentimental Analysis

# Objective

Financial news headlines are a fertile source of NLP data, especially when it comes to predicting how a stock will perform. Frequently, this is done via sentiment analysis, an NLP task that buckets phrases into positive, negative, and neutral.

This is a data of financial market top 25 news for the day and task is to train and predict model for overall sentimental analysis

# Import Library

In [1]:
import pandas as pd

In [2]:
import numpy as np

# Import dataset

In [4]:
df = pd.read_csv('C:\\Users\\PRIYESH\\OneDrive\\Desktop\\datasets\\financial market news.csv')

In [5]:
df.head()

Date  Label                                             News 1  \
0  01-01-2010      0           McIlroy's men catch cold from Gudjonsson   
1  02-01-2010      0               Warning from history points to crash   
2  03-01-2010      0      Comment: Why Israel's peaceniks feel betrayed   
3  04-01-2010      1              750,000-a-goal Weah aims parting shot   
4  05-01-2010      1  Leeds arrive in Turkey to the silence of the fans   

                                       News 2  \
0                       Obituary: Brian Walsh   
1              Investors flee to dollar haven   
2  Court deals blow to seizure of drug assets   
3            Newcastle pay for Fletcher years   
4     One woman's vision offers loan lifeline   

                                            News 3  \
0       Workplace blues leave employers in the red   
1                      Banks and tobacco in favour   
2                       An ideal target for spooks   
3  Brown sent to the stands for Scotland qualifier   
4          Working Lives: How world leaders worked   

                                        News 4  \
0                     Classical review: Rattle   
1                        Review: Llama Farmers   
2  World steps between two sides intent on war   
3         Tourists wary of breaking new ground   
4           Working Lives: Tricks of the trade   

                                              News 5  \
0                     Dance review: Merce Cunningham   
1                       War jitters lead to sell-off   
2                       What the region's papers say   
3              Canary Wharf climbs into the FTSE 100   
4  Working Lives: six-hour days, long lunches and...   

                                         News 6  \
0  Genetic tests to be used in setting premiums   
1                    Your not-so-secret history   
2           Comment: Fear and rage in Palestine   
3                           Review: Bill Bailey   
4                        Pop review: We Love UK   

                                          News 7  \
0                        Opera review: La Boh me   
1                  Review: The Northern Sinfonia   
2  Poverty and resentment fuels Palestinian fury   
3                              Review: Classical   
4               World music review: Marisa Monte   

                                        News 8  ...  \
0                   Pop review: Britney Spears  ...   
1                             Review: Hysteria  ...   
2  Republican feud fear as dissident is killed  ...   
3              Review: New Contemporaries 2000  ...   
4             Art review:  Hollingsworth/Heyer  ...   

                                News 16                              News 17  \
0                 Finland 0 - 0 England                   Healy a marked man   
1  Why Wenger will stick to his Gunners  Out of luck England hit rock bottom   
2             FTSE goes upwardly mobile             At this price?  BP Amoco   
3         More cash on way for counties      Cairns carries Kiwis to victory   
4          Duisenberg in double trouble           Pru to cut pension charges   

                                             News 18  \
0                     Happy birthday Harpers & Queen   
1                         Wilkinson out of his depth   
2                                            Go fish   
3  Year of Blanchflower's flourish when Spurs sto...   
4                            Art review: Paul Graham   

                                             News 19  \
0  Win unlimited access to the Raindance film fes...   
1                   Kinsella sparks Irish power play   
2         Bosnian Serb blows himself up to evade law   
3      New direct approach brings only pay-per-blues   
4               Shearer shot sparks Boro humiliation   

                                             News 20  \
0  Labour pledges  800m to bridge north-south divide   
1                    Brown banished as Scots rebound   
2                       Or

In [7]:
df.tail()

Date  Label                                             News 1  \
4096  20-03-2021      0  Barclays and RBS shares suspended from trading...   
4097  21-03-2021      1  2,500 Scientists To Australia: If You Want To ...   
4098  22-03-2021      1                   Explosion At Airport In Istanbul   
4099  23-03-2021      1  Jamaica proposes marijuana dispensers for tour...   
4100  24-03-2021      1  A 117-year-old woman in Mexico City finally re...   

                                                 News 2  \
4096  Pope says Church should ask forgiveness from g...   
4097  The personal details of 112,000 French police ...   
4098  Yemeni former president: Terrorism is the offs...   
4099  Stephen Hawking says pollution and 'stupidity'...   
4100   IMF chief backs Athens as permanent Olympic host   

                                                 News 3  \
4096  Poland 'shocked' by xenophobic abuse of Poles ...   
4097  S&amp;P cuts United Kingdom sovereign credit r...   
4098  UK must accept freedom of movement to access E...   
4099  Boris Johnson says he will not run for Tory pa...   
4100  The president of France says if Brexit won, so...   

                                                 News 4  \
4096  There will be no second referendum, cabinet ag...   
4097                Huge helium deposit found in Africa   
4098  Devastated: scientists too late to captive bre...   
4099  Six gay men in Ivory Coast were abused and for...   
4100  British Man Who Must Give Police 24 Hours' Not...   

                                                 News 5  \
4096      Scotland welcome to join EU, Merkel ally says   
4097  CEO of the South African state broadcaster qui...   
4098  British Labor Party leader Jeremy Corbyn loses...   
4099  Switzerland denies citizenship to Muslim immig...   
4100  100+ Nobel laureates urge Greenpeace to stop o...   

                                                 News 6  \
4096  Sterling dips below Friday's 31-year low amid ...   
4097  Brexit cost investors $2 trillion, the worst o...   
4098  A Muslim Shop in the UK Was Just Firebombed Wh...   
4099  Palestinian terrorist stabs israeli teen girl ...   
4100  Brazil: Huge spike in number of police killing...   

                                                 News 7  \
4096  No negative news about South African President...   
4097  Hong Kong democracy activists call for return ...   
4098  Mexican Authorities Sexually Torture Women in ...   
4099  Puerto Rico will default on $1 billion of debt...   
4100  Austria's highest court annuls presidential el...   

                                                 News 8  ...  \
4096  Surge in Hate Crimes in the U.K. Following U.K...  ...   
4097  Brexit: Iceland president says UK can join 'tr...  ...   
4098            UK shares and pound continue to recover  ...   
4099  Republic of Ireland fans to be awarded medal f...  ...   
4100  Facebook wins privacy case, can track any Belg...  ...   

                                                News 16  \
4096  German lawyers to probe Erdogan over alleged w...   
4097  US, Canada and Mexico pledge 50% of power from...   
4098  Escape Tunnel, Dug by Hand, Is Found at Holoca...   
4099  Googles free wifi at Indian railway stations i...   
4100  The United States has placed Myanmar, Uzbekist...   

                                                News 17  \
4096  Boris Johnson says the UK will continue to "in...   
4097  There is increasing evidence that Australia is...   
4098  The land under Beijing is sinking by as much a...   
4099  Mounting evidence suggests 'hobbits' were wipe...   
4100  S&amp;P revises European Union credit rating t...   

                                                News 18  \
4096  Richard Branson is calling on the UK governmen...   
4097  Richard Branson, the founder of Virgin Group, ...   
4098  Car bomb and Anti-Islamic attack on Mosque in ...   
4099  The men who carried out Tuesday's terror attac...   
4100  India gets $1 bil

# Describe Data

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4101 entries, 0 to 4100
Data columns (total 27 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   Date     4101 non-null   object
 1   Label    4101 non-null   int64 
 2   News 1   4101 non-null   object
 3   News 2   4101 non-null   object
 4   News 3   4101 non-null   object
 5   News 4   4101 non-null   object
 6   News 5   4101 non-null   object
 7   News 6   4101 non-null   object
 8   News 7   4101 non-null   object
 9   News 8   4101 non-null   object
 10  News 9   4101 non-null   object
 11  News 10  4101 non-null   object
 12  News 11  4101 non-null   object
 13  News 12  4101 non-null   object
 14  News 13  4101 non-null   object
 15  News 14  4101 non-null   object
 16  News 15  4101 non-null   object
 17  News 16  4101 non-null   object
 18  News 17  4101 non-null   object
 19  News 18  4101 non-null   object
 20  News 19  4101 non-null   object
 21  News 20  4101 non-null   object
 22  

In [13]:
df.describe()

Label
count  4101.000000
mean      0.528164
std       0.499267
min       0.000000
25%       0.000000
50%       1.000000
75%       1.000000
max       1.000000

In [14]:
df.shape

(4101, 27)

In [15]:
df.columns

Index(['Date', 'Label', 'News 1', 'News 2', 'News 3', 'News 4', 'News 5',
       'News 6', 'News 7', 'News 8', 'News 9', 'News 10', 'News 11', 'News 12',
       'News 13', 'News 14', 'News 15', 'News 16', 'News 17', 'News 18',
       'News 19', 'News 20', 'News 21', 'News 22', 'News 23', 'News 24',
       'News 25'],
      dtype='object')

# Data preprocessing

# Feature Selection

In [33]:
' '.join(str(x) for x in df.iloc[1,2:27])

"Warning from history points to crash Investors flee to dollar haven Banks and tobacco in favour Review: Llama Farmers War jitters lead to sell-off Your not-so-secret history Review: The Northern Sinfonia Review: Hysteria Review: The Guardsman Opera: The Marriage of Figaro Review: The Turk in Italy Deutsche spells out its plans for diversification Traders' panic sends oil prices skyward TV sport chief leaves home over romance Leader: Hi-tech twitch Why Wenger will stick to his Gunners Out of luck England hit rock bottom Wilkinson out of his depth Kinsella sparks Irish power play Brown banished as Scots rebound Battling Wales cling to lifeline Ehiogu close to sealing Boro move Man-to-man marking Match stats French referee at centre of storm is no stranger to controversy"

In [34]:
df.index

RangeIndex(start=0, stop=4101, step=1)

In [35]:
len(df.index)

4101

In [36]:
news = []
for row in range(0, len(df.index)):
    news.append(' '.join(str(x) for x in df.iloc[row, 2:27]))

In [37]:
type(news)

list

In [38]:
news[0]

"McIlroy's men catch cold from Gudjonsson Obituary: Brian Walsh Workplace blues leave employers in the red Classical review: Rattle Dance review: Merce Cunningham Genetic tests to be used in setting premiums Opera review: La Boh me Pop review: Britney Spears Theatre review: The Circle Wales face a fraught night Under-21  round-up Smith off to blot his copybook Finns taking the mickey Praise wasted as Brown studies injury options Ireland wary of minnows Finland 0 - 0 England Healy a marked man Happy birthday Harpers & Queen Win unlimited access to the Raindance film festival Labour pledges  800m to bridge north-south divide Wales: Lib-Lab pact firm despite resignation Donald Dewar Regenerating homes  regenerates well-being in people Win  100 worth of underwear TV guide: Random views"

In [39]:
X = news

In [40]:
type(X)

list

# Text conversion to bag of words

In [41]:
from sklearn.feature_extraction.text import CountVectorizer

In [42]:
cv = CountVectorizer(lowercase = True, ngram_range = (1,1))

In [43]:
X = cv.fit_transform(X)

In [44]:
X.shape

(4101, 48463)

In [45]:
y = df['Label']

In [46]:
y.shape

(4101,)

# Train Test Split dataset

In [47]:
from sklearn.model_selection import train_test_split

In [49]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, stratify = y, random_state = 2520)

In [50]:
from sklearn.ensemble import RandomForestClassifier

In [51]:
rf = RandomForestClassifier(n_estimators = 200)

In [52]:
rf.fit(X_train, y_train)

RandomForestClassifier(n_estimators=200)

In [53]:
y_pred = rf.predict(X_test)

In [55]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

In [56]:
confusion_matrix(y_test, y_pred)

array([[126, 455],
       [149, 501]], dtype=int64)

In [57]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.46      0.22      0.29       581
           1       0.52      0.77      0.62       650

    accuracy                           0.51      1231
   macro avg       0.49      0.49      0.46      1231
weighted avg       0.49      0.51      0.47      1231



# Explanation 

This result is based on a binary classification task, where there are two classes: 0 and 1. Let's break down the key metrics:

Precision:

Precision is a measure of how many of the predicted positive instances were actually true positives. In this case, for class 0, the precision is 0.46, which means that out of all the instances predicted as class 0, 46% of them were actually class 0. For class 1, the precision is 0.52, indicating that 52% of the instances predicted as class 1 were truly class 1.

Recall:

Recall (also called True Positive Rate or Sensitivity) is a measure of how many of the true positive instances were correctly predicted. In this case, for class 0, the recall is 0.22, which means that only 22% of all the actual class 0 instances were correctly predicted as class 0. For class 1, the recall is 0.77, indicating that 77% of the actual class 1 instances were correctly predicted as class 1.

F1-Score:

The F1-score is a metric that combines precision and recall into a single value. It's useful when you want to balance both false positives and false negatives. The formula for the F1-score is 2 * (precision * recall) / (precision + recall).
For class 0, the F1-score is 0.29, and for class 1, it's 0.62. Higher F1-scores indicate better balance between precision and recall.

Support:

Support represents the number of instances in each class. In this case, there are 581 instances of class 0 and 650 instances of class 1 in the dataset.

Accuracy:

Accuracy is a measure of overall correctness and is not class-specific. It calculates the proportion of correctly predicted instances out of all instances. In this case, the overall accuracy is 0.51 or 51%.

Macro Average and Weighted Average:

These are average values calculated across both classes.
The macro average calculates the average of precision, recall, and F1-score independently for each class and then takes their arithmetic mean.
The weighted average considers the contributions of each class based on their support (the number of instances in each class). It's weighted by the number of instances in each class, so classes with more instances have a greater impact on the average.

# Summary

In summary, your model has higher precision and recall for class 1, indicating that it's better at correctly identifying instances of class 1. However, it has lower precision and recall for class 0. The F1-scores provide a balanced view of the model's performance, with class 1 having a higher F1-score, suggesting better overall performance for that class. Overall accuracy is relatively low at 51%, indicating that the model's performance may need improvement, especially in terms of precision and recall for class 0.